// grav.client/Common/TestDoubles/json-response-stubs.ts
import * as _existsJsonResponse from "../Assets/Json/grav.exists.json";
import * as _existsJsonResponseMultiple from "../Assets/Json/grav.exists.multiple.json";
import * as _addressesJsonResponse from "../Assets/Json/grav.addresses.json";
import * as _userImagesJsonResponse from "../Assets/Json/grav.userImages.json";
import * as _userImagesNoneJsonResponse from "../Assets/Json/grav.userImages.none.json";
import * as _userImagesSingleJsonResponse from "../Assets/Json/grav.userImages.single.json";
import * as _saveImageResponse from "../Assets/Json/grav.saveImage.json";
import * as _useUserImageJsonResponse from "../Assets/Json/grav.useUserImage.json";
import * as _removeImageJsonResponse from "../Assets/Json/grav.removeImage.json";
import * as _testJsonResponse from "../Assets/Json/grav.test.json";

export const existsJsonResponse = _existsJsonResponse;
export const existsJsonResponseMultiple = _existsJsonResponseMultiple;
export const addressesJsonResponse = _addressesJsonResponse;
export const userImagesJsonResponse = _userImagesJsonResponse;
export const userImagesNoneJsonResponse = _userImagesNoneJsonResponse;
export const userImagesSingleJsonResponse = _userImagesSingleJsonResponse;
export const saveImageResponse = _saveImageResponse;
export const useUserImageJsonResponse = _useUserImageJsonResponse;
export const removeImageJsonResponse = _removeImageJsonResponse;
export const testJsonResponse = _testJsonResponse;
