// grav.client/Release/Common/TestDoubles/json-response-stubs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.testJsonResponse = exports.removeImageJsonResponse = exports.useUserImageJsonResponse = exports.saveImageResponse = exports.userImagesSingleJsonResponse = exports.userImagesNoneJsonResponse = exports.userImagesJsonResponse = exports.addressesJsonResponse = exports.existsJsonResponseMultiple = exports.existsJsonResponse = void 0;
const _existsJsonResponse = __importStar(require("../Assets/Json/grav.exists.json"));
const _existsJsonResponseMultiple = __importStar(require("../Assets/Json/grav.exists.multiple.json"));
const _addressesJsonResponse = __importStar(require("../Assets/Json/grav.addresses.json"));
const _userImagesJsonResponse = __importStar(require("../Assets/Json/grav.userImages.json"));
const _userImagesNoneJsonResponse = __importStar(require("../Assets/Json/grav.userImages.none.json"));
const _userImagesSingleJsonResponse = __importStar(require("../Assets/Json/grav.userImages.single.json"));
const _saveImageResponse = __importStar(require("../Assets/Json/grav.saveImage.json"));
const _useUserImageJsonResponse = __importStar(require("../Assets/Json/grav.useUserImage.json"));
const _removeImageJsonResponse = __importStar(require("../Assets/Json/grav.removeImage.json"));
const _testJsonResponse = __importStar(require("../Assets/Json/grav.test.json"));
exports.existsJsonResponse = _existsJsonResponse;
exports.existsJsonResponseMultiple = _existsJsonResponseMultiple;
exports.addressesJsonResponse = _addressesJsonResponse;
exports.userImagesJsonResponse = _userImagesJsonResponse;
exports.userImagesNoneJsonResponse = _userImagesNoneJsonResponse;
exports.userImagesSingleJsonResponse = _userImagesSingleJsonResponse;
exports.saveImageResponse = _saveImageResponse;
exports.useUserImageJsonResponse = _useUserImageJsonResponse;
exports.removeImageJsonResponse = _removeImageJsonResponse;
exports.testJsonResponse = _testJsonResponse;
