// grav.client/Release/Infrastructure/http-shim.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpShim = exports.origin = void 0;
require("isomorphic-unfetch");
exports.origin = "https://secure.gravatar.com";
class HttpShim {
    constructor(emailHash) {
        this.endpoint = `${exports.origin}/xmlrpc?user=${emailHash}`;
    }
    rpc(message) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetch(this.endpoint, {
                method: "POST",
                headers: { "content-type": "text/xml" },
                body: message,
            });
        });
    }
}
exports.HttpShim = HttpShim;
